nucleus/scene.py

```diff
@@ -0,0 +1,78 @@
+from dataclasses import dataclass
+from typing import Optional, Dict, List, Set
+from enum import Enum
+from .annotation import Point3D
+from .utils import flatten
+
+
+class DatasetItemType(Enum):
+    IMAGE = "image"
+    POINTCLOUD = "pointcloud"
+    VIDEO = "video"
+
+
+@dataclass
+class Quaternion:
+    x: float
+    y: float
+    z: float
+    w: float
+
+
+@dataclass
+class CameraParams:
+    position: Point3D
+    heading: Quaternion
+    fx: float
+    fy: float
+    cx: float
+    cy: float
+
+
+@dataclass
+class SceneDatasetItem:
+    url: str
+    type: DatasetItemType
+    reference_id: Optional[str] = None
+    metadata: Optional[dict] = None
+    camera_params: Optional[CameraParams] = None
+
+
+@dataclass
+class Frame:
+    items: Dict[str, SceneDatasetItem]
+
+    def __post_init__(self):
+        for key, value in self.items.items():
+            assert isinstance(key, str), "All keys must be names of sensors"
+            assert isinstance(
+                value, SceneDatasetItem
+            ), "All values must be SceneDatasetItems"
+
+
+@dataclass
+class Scene:
+    frames: List[Frame]
+    reference_id: str
+    metadata: Optional[dict] = None
+
+
+@dataclass
+class LidarScene(Scene):
+    def __post_init__(self):
+        # do validation here for lidar scene
+        lidar_sources = flatten(
```

Comment:
I think a user would actually construct the scene over time i.e. initialize a scene object, call scene.add_item() repeatedly (or something roughly along those lines) and then upload via API.  I therefore think validation is something that should happen on upload rather than post-init.

Reply:
moved to separate function - will call before upload